File: 数据结构与算法/树/btree.js
/*

二叉树

*/

/** 二叉树节点 */
class BNode {
    constructor(val = 0) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

/**
 * 先序遍历
 * @param root 二叉树根节点
 */
function preorder(root) {
    if (root) {
        process.stdout.write(root.val + ' ');
        preorder(root.left);
        preorder(root.right);
    }
}

/**
 * 中序遍历
 * @param root 二叉树根节点
 */
function inorder(root) {
    if (root) {
        inorder(root.left);
        process.stdout.write(root.val + ' ');
        inorder(root.right);
    }
}

/**
 * 后序遍历
 * @param root 二叉树根节点
 */
function postorder(root) {
    if (root) {
        postorder(root.left);
        postorder(root.right);
        process.stdout.write(root.val + ' ');
    }
}

/**
 * 层序遍历
 * @param root 二叉树根节点
 */
function levelorder(root) {
    if (!root)
        return;

    let que = new Array();
    que.push(root);
    while (que.length) {
        let n = que.shift();
        process.stdout.write(n.val + ' ');
        if (n.left)
            que.push(n.left);
        if (n.right)
            que.push(n.right);
    }
}

function main() {

    let root = new BNode(1);
    root.left = new BNode(2);
    root.right = new BNode(3);
    root.left.left = new BNode(4);
    root.right.left = new BNode(5);

    /*
    
            1
           / \
          2   3
         /   /
        4   5  

    */

    preorder(root);
    console.log();

    inorder(root);
    console.log();

    postorder(root);
    console.log();

    levelorder(root);
    console.log();
}

main();

File: 数据结构与算法/树/btree.py
'''

二叉树

'''


class BTNode:
    '''
    二叉树节点
    '''
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def preorder(root: BTNode):
    '''
    先序遍历
    '''
    if root:
        print(root.val, end=' ')
        preorder(root.left)
        preorder(root.right)


def inorder(root: BTNode):
    '''
    中序遍历
    '''
    if root:
        inorder(root.left)
        print(root.val, end=' ')
        inorder(root.right)


def postorder(root: BTNode):
    '''
    后序遍历
    '''
    if root:
        postorder(root.left)
        postorder(root.right)
        print(root.val, end=' ')


def levelorder(root: BTNode):
    '''
    层序遍历
    '''
    if not root:
        return
    st = [root]
    while st:
        p = st.pop(0)
        print(p.val, end=' ')
        if p.left:
            st.append(p.left)
        if p.right:
            st.append(p.right)


def preorder_c(root: BTNode):
    '''
    非递归版先序遍历
    '''
    st = []
    while root or st:
        if root:
            print(root.val, end=' ')
            st.append(root)
            root = root.left
        else:
            root = st.pop().right


def inorder_c(root: BTNode):
    '''
    非递归版中序遍历
    '''
    st = []
    while root or st:
        if root:
            st.append(root)
            root = root.left
        else:
            root = st.pop()
            print(root.val, end=' ')
            root = root.right


def postorder_c(root: BTNode):
    '''
    非递归版后序遍历
    '''
    r = None
    st = []
    while root or st:
        if root:
            st.append(root)
            root = root.left
        else:
            root = st[-1]
            if root.right and r != root.right:
                root = root.right
            else:
                st.pop()
                print(root.val, end=' ')
                r = root
                root = None


if __name__ == "__main__":
    '''
        1
       / \
      2   3
     /     \
    4       5
    '''

    root = BTNode(1)
    root.left = BTNode(2)
    root.right = BTNode(3)
    root.left.left = BTNode(4)
    root.right.right = BTNode(5)

    preorder(root)
    print()
    inorder(root)
    print()
    postorder(root)
    print()
    levelorder(root)
    print()

    print()

    preorder_c(root)
    print()
    inorder_c(root)
    print()
    postorder_c(root)
    print()
